indico/modules/events/registration/models/registrations.py

```diff
@@ -58,9 +58,9 @@ class Registration(db.Model):
     __table_args__ = (db.CheckConstraint('email = lower(email)', 'lowercase_email'),
                       db.Index(None, 'friendly_id', 'event_id', unique=True),
                       db.Index(None, 'registration_form_id', 'user_id', unique=True,
-                               postgresql_where=db.text('state NOT IN (3, 4)')),
+                               postgresql_where=db.text('NOT is_deleted OR (state NOT IN (3, 4))')),
```

Comment:
no need for the parentheses around the `state not in` part (but they don't hurt either).

but you need `AND` instead of `OR` here.